### PR TITLE
[BEAM-8335] Implemented Capture Size limitation

### PR DESCRIPTION
**Pull request by KevinGG**
1. Capture cache size is calculated as a property from StreamingCacheSinks and
   should be implemented by any other sink implementations in the future.
2. For each capture cache, if it doesn't exist, the size is 0.
3. If the cache manager doesn't exist or doesn't support capture_size,
   the capture size configured in ib.options would not take effect as if
   capture cache size stays 0.
4. Captured data from sources are stored like intermediate PCollections.
   The capture cache size does not include intermediate PCollections'
   disk usages.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by KevinGG** [comment]
R: @rohdesamuel 
R: @aaltay 

Data captured from sources are stored in cache just like intermediate PCollections that are assigned to variables.
The capture_size limit is only applied to disk usage of data captured from sources.

The implementation of getting a capture cache file's size
`os.stat(self._path).st_size`

The implementation of summing up all capture cache file's sizes
`sum([sink.size_in_bytes for _, sink in self._capture_sinks.items()])`

Everything else is wiring, logging and testing.

File: sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py [hunk 1]
```diff
@@ -63,6 +63,14 @@ def path(self):
     """Returns the path the sink leads to."""
     return self._path
 
+  @property
+  def size_in_bytes(self):
+    """Returns the space usage in bytes of the sink."""
+    try:
+      return os.stat(self._path).st_size
+    except:
+      return 0
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py]: Maybe log here?

File: sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py [hunk 2]
```diff
@@ -63,6 +63,14 @@ def path(self):
     """Returns the path the sink leads to."""
     return self._path
 
+  @property
+  def size_in_bytes(self):
+    """Returns the space usage in bytes of the sink."""
+    try:
+      return os.stat(self._path).st_size
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py]: Does this work in all OSes?

File: sdks/python/apache_beam/runners/interactive/interactive_beam.py [hunk 3]
```diff
@@ -88,7 +88,22 @@ def capture_duration(self, value):
     """
     self.capture_control._capture_duration = value
 
-  # TODO(BEAM-8335): add capture_size options when they are supported.
+  @property
+  def capture_size(self):
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/interactive_beam.py]: Shoulds this have "limit" or "max" in its name, or something to indicate that this is an upper limit.

File: sdks/python/apache_beam/runners/interactive/background_caching_job.py [hunk 4]
```diff
@@ -259,21 +269,31 @@ def is_source_to_cache_changed(
   # The computation of extract_unbounded_source_signature is expensive, track on
   # change by default.
   if is_changed and update_cached_source_signature:
-    if ie.current_env().options.enable_capture_replay:
-      # TODO(BEAM-8335): display rather than logging when is_in_notebook.
+    options = ie.current_env().options
+    # No info needed when capture replay is disabled.
+    if options.enable_capture_replay:
+
+      def sizeof_fmt(num, suffix='B'):
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/background_caching_job.py]: Should this be moved inside the if below? Is it used anywhere else?

File: sdks/python/apache_beam/runners/interactive/background_caching_job.py [hunk 5]
```diff
@@ -202,10 +202,20 @@ def has_source_to_cache(user_pipeline):
...
+        job = ie.current_env().get_background_caching_job(user_pipeline)
+        is_done = job and job.is_done()
+        cache_changed = is_source_to_cache_changed(
+            user_pipeline, update_cached_source_signature=False)
+        return is_done and not cache_changed
+
+      ie.current_env().set_cache_manager(
+          streaming_cache.StreamingCache(
+              ie.current_env().cache_manager()._cache_dir,
+              is_cache_complete=is_cache_complete,
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/interactive/background_caching_job.py]: Do you mean to pass `is_cache_complete()` or `is_cache_complete` ?

Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/background_caching_job.py]: `is_cache_complete` is a function required by StreamingCache, not the return value of the function.

File: sdks/python/apache_beam/runners/interactive/background_caching_job.py (hunk 4 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/background_caching_job.py]: No, it's not used anywhere else. Moving it.

File: sdks/python/apache_beam/runners/interactive/interactive_beam.py (hunk 3 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/interactive_beam.py]: Renaming it to `capture_size_limit`.

File: sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py (hunk 2 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py]: It works. Though in some versions of some OSes, some of the fields of the returned stat might be dummy values.

File: sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py (hunk 1 above)
Comment (KevinGG) [reply sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py]: Could add a debug level logging.

**Comment by KevinGG** [comment]
Waiting on merge of https://github.com/apache/beam/pull/11005, then I'll resolve the merge conflicts.

**Comment by KevinGG** [comment]
Resolved merge conflicts and force pushed.

**Comment by KevinGG** [comment]
Discarded outdated changes from dependency PRs and squashed the empty patch commits.

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
retest this please

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Run Python PreCommit

**Comment by aaltay** [comment]
Run Python PreCommit

**Pull request merged**